include/mbgl/style/expression/match.hpp

```diff
@@ -0,0 +1,277 @@
+#pragma once
+
+#include <mbgl/style/expression/expression.hpp>
+#include <mbgl/style/expression/parsing_context.hpp>
+#include <mbgl/style/conversion.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+using InputType = variant<int64_t, std::string>;
+using MatchKey = variant<InputType, std::vector<InputType>>;
+
+template <typename T>
+class TypedMatch : public TypedExpression {
+public:
+    TypedMatch(std::unique_ptr<TypedExpression> input_,
+               std::vector<std::pair<MatchKey, std::unique_ptr<TypedExpression>>> cases_,
+               std::unique_ptr<TypedExpression> otherwise_
+    ) : TypedExpression(otherwise_->getType()),
+        input(std::move(input_)),
+        otherwise(std::move(otherwise_))
+    {
+        for(auto& pair : cases_) {
+            pair.first.match(
+                [&](const InputType& key) {
+                    cases.emplace(key.get<T>(), std::move(pair.second));
+                },
+                [&](const std::vector<InputType>& keys) {
+                    std::shared_ptr<TypedExpression> output = std::move(pair.second);
+                    for (const auto& key : keys) {
+                        cases.emplace(key.get<T>(), output);
+                    }
+                }
+            );
+        }
+    }
+    
+    bool isFeatureConstant() const override {
+        if (!input->isFeatureConstant()) { return false; }
+        if (!otherwise->isFeatureConstant()) { return false; }
+        for (const auto& pair : cases) {
+            if (!pair.second->isFeatureConstant()) { return false; }
+        }
+        return true;
+    }
+
+    bool isZoomConstant() const override {
+        if (!input->isZoomConstant()) { return false; }
+        if (!otherwise->isZoomConstant()) { return false; }
+        for (const auto& pair : cases) {
+            if (!pair.second->isZoomConstant()) { return false; }
+        }
+        return true;
+    }
+    
+    EvaluationResult evaluate(const EvaluationParameters& params) const override {
+        const auto& inputValue = evaluateInput(params);
+        if (!inputValue) {
+            return inputValue.error();
+        }
+        if (cases.find(*inputValue) == cases.end()) {
+            return otherwise->evaluate(params);
+        }
+        return cases.at(*inputValue)->evaluate(params);
```

Comment:
Use an iterator variable rather than finding `*inputValue` twice (once with `find` and once with `at`).